src/reason_pprint_ast.ml

```diff
@@ -2996,94 +3092,142 @@ class printer  ()= object(self:'self)
-      | Pexp_apply (e, ls) -> (
-          match (view_fixity_of_exp e, ls) with
-            | ((`Infix infixStr), [("", leftExpr); ("", rightExpr)]) -> (
-                match self#consecutiveInfixApplicationItemsRecurse x with
-                  | (None, _) -> None
-                  | (Some infix, lst) -> Some(infix, lst)
-              )
-            | _ -> None
-        )
-      | _  -> None
-
-  (**
-     TODO: Configure the optional ability to print the *minimum* number of
-     parens. It's simply a matter of changing [higherPrecedenceThan] to
-     [higherOrEqualPrecedenceThan].
-   *)
-  method consecutiveInfixApplicationItemsRecurse x =
-    let rightSideContribution infixStr origRight (rightRecurseOp, rightRecurseList) =
-      match rightRecurseOp with
-        | (Some (s, _)) ->
-          if not (s = infixStr) then
-            (if (higherPrecedenceThan s infixStr) then [self#expression origRight] else [self#simple_expression origRight])
-          else
-            (if isRightAssociative infixStr then rightRecurseList else [self#simple_expression origRight])
-        | None -> rightRecurseList (* Will only ever be len 1 *)
-    in
-    let leftSideContribution infixStr origLeft (leftRecurseOp, leftRecurseList) =
-      match leftRecurseOp with
-        | (Some (s, _)) ->
-          if not (s = infixStr) then
-            (if (higherPrecedenceThan s infixStr) then [self#expression origLeft] else [self#simple_expression origLeft])
-          else
-            (if isLeftAssociative infixStr then leftRecurseList else [self#simple_expression origLeft])
-        | None -> leftRecurseList (* Will only ever be len 1 *)
-    in
-    match x.pexp_desc with
-      | Pexp_apply (e, ls) -> (
-        match (view_fixity_of_exp e, ls) with
-          | (`Infix infixStr, [("", leftExpr); ("", rightExpr)]) ->
-            let rightRecurse = self#consecutiveInfixApplicationItemsRecurse rightExpr in
-            let leftRecurse = self#consecutiveInfixApplicationItemsRecurse leftExpr in
-            let rightItms = rightSideContribution infixStr rightExpr rightRecurse in
-            let leftItms = leftSideContribution infixStr leftExpr leftRecurse in
-            ((Some (infixStr, e.pexp_loc)), List.concat [leftItms; rightItms])
-          (* Again, it's reall good that this only ever is caught at the
-             recursive step - otherwise we'd loop infinitely. *)
-          | (`Infix infixStr, lbls) -> (None, [self#expression x])
-          | (`Prefix s, _) -> (None, [self#simple_expression x])
-          (* We know that this can only ever occur at a recursive call - the
-             first call into this would have been infix. There is overlap with
-             this and the simple_enough_to_be_separated_by_infix case. Either
-             path would work. *)
-          | (`Normal, _) -> (None, [self#expression x])
-        )
-      | _ -> (None, [self#simple_enough_to_be_separated_by_infix x])
+    | Pexp_apply (e, ls) -> (
```

Comment:
Unparsing of function application, including the various potential forms (such as infix/prefix/sugary setting of array fields etc).